internal/server/iito.go

```diff
@@ -48,56 +48,103 @@ func (s *SOMASServer) runGiftSession() error {
 	return nil
 }
 
-func (s *SOMASServer) getGiftRequests() shared.GiftDict {
-	giftRequestDict := shared.GiftDict{}
-	for id, client := range s.clientMap {
-		giftRequestDict[id] = client.RequestGift()
+func (s *SOMASServer) sanitiseTeamGiftRequests(requests shared.GiftRequestDict, thisTeam shared.ClientID) (shared.GiftRequestDict, error) {
+	for team, request := range requests {
+		// Delete the request if it's to yourself or a dead team.
+		if s.gameState.ClientInfos[team].LifeStatus == shared.Dead || team == thisTeam || request == 0 {
+			delete(requests, team)
+		}
+	}
+
+	// TODO: Maybe return some kind of helpful message if any of the above cases break?
+	return requests, nil
+}
+
+// GetGiftRequests collects a map of gift requests from an individual client, for all clients, in a map
+func (s *SOMASServer) getGiftRequests() map[shared.ClientID]shared.GiftRequestDict {
+	totalRequests := map[shared.ClientID]shared.GiftRequestDict{}
+	for _, id := range getNonDeadClientIDs(s.gameState.ClientInfos) {
+		totalRequests[id], _ = s.sanitiseTeamGiftRequests(s.clientMap[id].GetGiftRequests(), id)
+
+		if len(totalRequests[id]) == 0 {
+			delete(totalRequests, id)
+		}
+	}
+	return totalRequests
+}
+
+// TODO: Return an error?
+func (s *SOMASServer) sanitiseTeamGiftOffers(offers shared.GiftOfferDict, thisTeam shared.ClientID) (shared.GiftOfferDict, error) {
+	for team, offer := range offers {
+		if s.gameState.ClientInfos[team].LifeStatus == shared.Dead || team == thisTeam || offer == 0 {
+			delete(offers, team)
+		}
 	}
-	return giftRequestDict
+	// TODO: Cap the last offer(s) so that the sum of offers doesn't exceed total resources
+	// TODO: Maybe return some kind of helpful message if any of the above cases break?
+	return offers, nil
 }
-func (s *SOMASServer) getGiftOffers(giftRequestDict shared.GiftDict) (map[shared.ClientID]shared.GiftDict, error) {
-	giftOfferDict := map[shared.ClientID]shared.GiftDict{}
-	var err error
-	for id, client := range s.clientMap {
-		giftOfferDict[id], err = client.OfferGifts(giftRequestDict)
+
+// getGiftOffers collects all responses from clients to their requests in a map
+func (s *SOMASServer) getGiftOffers(totalRequests map[shared.ClientID]shared.GiftRequestDict) (map[shared.ClientID]shared.GiftOfferDict, error) {
+	totalOffers := map[shared.ClientID]shared.GiftOfferDict{}
+	for _, thisTeam := range getNonDeadClientIDs(s.gameState.ClientInfos) {
+		// Gather all the requests made to this team
+		requestsToThisTeam := shared.GiftRequestDict{}
+		for fromTeam, indivRequests := range totalRequests {
+			requestsToThisTeam[fromTeam] = indivRequests[thisTeam]
```

Comment:
If  fromTeam does not request anything from thisTeam, then the requestsToThisTeam gets a default initialised GiftRequest which is 0. This is then passed to the clients. So essentially every client gets a request with a value of zero from the teams who requested nothing, which somewhat negates the purpose of the previous sanitisation. 

A similar issue is found on line 128 and on 145

Reply:
Nice catch. Didn't know invalid accesses are returned with default values. Fixed with the comma, ok pattern.